### ISSUE TITLE
[Bug] Template variables with dependencies get loaded multiple times

### ISSUE THREAD

**Issue by alin-amana** [grafana/grafana#9185]
Using Grafana 4.4.3 with Prometheus on Ubuntu.

I have a dashboard with 2 template variables that depend on one another, set to refresh on time range changes:
- `$_env = label_values(up, env)`
- `$_job = label_values(up{env="$_env"}, job)`

When the dashboard reloads (either periodically, as configured, or when I click Grafana's reload button in the top right) I would expect to see 2 `/api/v1/series` requests triggered, one for each variable.

Instead, these are the requests I see:
![Grafana 2 template variables](https://image.ibb.co/idAM4F/grafana_template_variables.png)

The first two are for each of the template variables (1), (2). Then comes a data request for the one graph I have on the dashboard (3). Followed by two more requests for the second template variable (4) (5). If I were to guess, request (4) is issued as soon as request (1) is complete (presumably using the values returned by (1), but they are the same as before and thus request (4) is absolutely identical, including start and end times, to request (2)). Still guessing here, I believe request (5) is issued once the whole dashboard has loaded. Again, completely identical to requests (2) and (4).

If I add a third template variable that depends on the second, also set to refresh on time range change -- `$_instance = label_values(up{env="$_env",job="$_job"}, instance)` -- I get an even wilder collection of requests on every single reload:
![Grafana 3 template variables](https://image.ibb.co/cYRdjF/grafana_3_template_variables.png)

They are, in order for:
- template variable 1 (TV1)
- template variable 2 (TV2)
- template variable 3 (TV3)
- graph data
- TV2
- TV3
- TV3
- TV3
- TV3
- TV2
- TV3
- TV3

All TV2 requests are identical. All TV3 requests are identical.

I have tried debugging this, but it's deep within Angular (IIRC) minified code and I have never done frontend development so I'm pretty much useless.

For the sake of completeness, here is the templating section of the dashboard JSON (the rest doesn't matter, this happens with an empty dashboard too):

```
  "templating": {
    "list": [
      {
        "allValue": null,
        "current": {
          "text": "prod",
          "value": "prod"
        },
        "datasource": "Prometheus",
        "hide": 0,
        "includeAll": false,
        "label": "Environment",
        "multi": false,
        "name": "_env",
        "options": [],
        "query": "label_values(up, env)",
        "refresh": 2,
        "regex": "",
        "sort": 0,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      },
      {
        "allValue": null,
        "current": {
          "text": "prometheus",
          "value": "prometheus"
        },
        "datasource": "Prometheus",
        "hide": 0,
        "includeAll": false,
        "label": "Job",
        "multi": false,
        "name": "_job",
        "options": [],
        "query": "label_values(up{env=\"$_env\"}, job)",
        "refresh": 2,
        "regex": "",
        "sort": 0,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      },
      {
        "allValue": null,
        "current": {
          "text": "9090",
          "value": "9090"
        },
        "datasource": "Prometheus",
        "hide": 0,
        "includeAll": false,
        "label": "Instance",
        "multi": false,
        "name": "_instance",
        "options": [],
        "query": "label_values(up{env=\"$_env\",job=\"$_job\"}, instance)",
        "refresh": 2,
        "regex": "/.*:(.*)/",
        "sort": 0,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      }
    ]
  },
```

**Comment by torkelo** [comment]
only set the root variable to refresh on time range change

**Comment by alin-amana** [comment]
Thanks for the tip, much appreciated.

However, it only marginally improved things, I am still seeing a number of problems. (I will use A for the root variable/dependency and B for the dependent variable below, it helps with brevity and consistence.)

1. I'm still seeing way too many `/api/v1/series` requests. To be more precise, there seem to be 2x the number of requests for B as for A. Meaning that in the 3 template variable dependency chain described above I'm seeing 1 request for `$_env`, 2 requests for `$_job` and 4 requests for `$_instance`.

1. This behavior is not documented. I've read through the template variable documentation and while it does mention that Grafana will "automatically refresh a variable when one of it’s containing variables change", there is no warning against explicitly setting up dependent variables (B in our case) to refresh on time range change. Automatic refresh is a really nice to have feature, but it should not negatively affect the case where B is set up to refresh on time range changes.

1. B is refreshed every time A is refreshed, not when the current value of A changes, not when the set of values for A change. Always.

1. I took the extreme approach when following your advice and set up B to never refresh, relying on the automatic refresh mechanism. This works just fine when I'm explicitly selecting a different value for A: B gets refreshed and the dashboard reflects both changes. However, on dashboard load A is refreshed (and new values, different from what's saved in the dashboard, are loaded) but B isn't (neither before the dashboard is laid out and loaded nor after), so the dashboard will reflect the value of A, but not B. In my particular case, on an all services dashboard that is supposed to show a singlestat panel for each service, the dashboard will have a row for each environment (prod, stage and test) but only an "All" job on each row (because my `$_job` variable has an All value). The jobs are never displayed, even when `$_job` (my B variable) is eventually refreshed and all values loaded into the dropbox, presumably because B is never supposed to refresh. If I then explicitly change the value of `$_env` (my A), i.e. I pick one of the environments (e.g. `prod`) then the dashboard is finally reloaded. But loaded and left on its own, doing periodic refreshes, it will never show what it's supposed to show. The workaround is to set B to refresh on dashboard reload, but it is not obvious.

**Comment by torkelo** [comment]
> I'm still seeing way too many /api/v1/series requests. To be more precise, there seem to be 2x the number of requests for B as for A. Meaning that in the 3 template variable dependency chain described above I'm seeing 1 request for $_env, 2 requests for $_job and 4 requests for $_instance.

Strange, seeing this as well. Will open & investigate!

**Comment by alin-amana** [comment]
Thanks for looking into it. As you can probably tell it's not exactly a critical bug, but it does add unnecessary load to Prometheus.

**Comment by torkelo** [comment]
this seems to only happen when you have a parent variable with Refresh "On Time Range Change", this seems to cause multiple updates of the child variables (that use the parent variable in its definition)

**Comment by nachopv** [comment]
Hi @torkelo 

We have been digging about this behaviour but still only updating the root value does not have a proper behaviour
Our deployment is isung Influx db datasources but regarding the chained variables update does not matter what datasource we are using but the policy on when the refresh is triggered. Also we have tested the behaviour both on Grafana 6 & 7 with the same result 

- Definitely using refresh "On Dashboard Load" only for the root variable prevents chained variables to be calculated multiple times
- However with this set up we have detected that when saharing a parametrized dashboard with other coleagues ( i.e. https://mygrafana/uid/name?var-v1=a&var-v2=b ) only the root variable is calculated but not the chained ones
- We tried to mitigate this with another setup
      - On Dashboard load only for root variable
      - On time range for al the chained variables

This setup solved the issue of sharing parametrized dashboard, but if you decide to zoom in the dashboard you can see that we do have again the issue of calculating multiple times the chained variables

As we do see that the patterns is related to the chaining level ( level2 executed 2 times, level 3 executed 3 times and so on ) we tried a final setup
      - On Dashboard load only for root variable ( level 1 )
      - On time range for al the chained variables ( level 2 )
      - Never for level 3+

This does not work as long as level+ variables are not calculated when sharing a parametrized dashboard link

Summing up we do think , it should be reviewed the logic of only using refresh policy on dashboard load for the rrot variable in order to trigger the calculation of the chained variables just once

**Comment by torkelo** [comment]
Maybe solved in upcoming 7.2 https://github.com/grafana/grafana/pull/27790
